backend/src/apiserver/metadata/metadata_store.go

```diff
@@ -0,0 +1,127 @@
+package metadata
+
+import (
+	"encoding/json"
+	"ml_metadata/metadata_store/mlmetadata"
+	mlpb "ml_metadata/proto/metadata_store_go_proto"
+	"strings"
+
+	argoWorkflow "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+	"github.com/kubeflow/pipelines/backend/src/common/util"
+)
+
+// Store encapsulates a ML Metadata store.
+type Store struct {
+	mlmdStore *mlmetadata.Store
+}
+
+// NewStore creates a new Store, using mlmdStore as the backing ML Metadata
+// store.
+func NewStore(mlmdStore *mlmetadata.Store) *Store {
+	return &Store{mlmdStore: mlmdStore}
+}
+
+// RecordOutputArtifacts records metadata on artifacts as parsed from the Argo
+// output parameters in currentManifest. storedManifest represents the currently
+// stored manifest for the run with id runID, and is used to ensure metadata is
+// recorded at most once per artifact.
+func (s *Store) RecordOutputArtifacts(runID, storedManifest, currentManifest string) error {
+	storedWorkflow := &argoWorkflow.Workflow{}
+	if err := json.Unmarshal([]byte(storedManifest), storedWorkflow); err != nil {
+		return util.NewInternalServerError(err, "unmarshaling workflow failed")
+	}
+
+	currentWorkflow := &argoWorkflow.Workflow{}
+	if err := json.Unmarshal([]byte(currentManifest), currentWorkflow); err != nil {
+		return util.NewInternalServerError(err, "unmarshaling workflow failed")
+	}
+
+	completed := make(map[string]bool)
+	for _, n := range storedWorkflow.Status.Nodes {
+		if n.Completed() {
+			completed[n.ID] = true
+		}
+	}
+
+	for _, n := range currentWorkflow.Status.Nodes {
+		if n.Completed() && !completed[n.ID] {
+			// Newly completed node. Record output ml-metadata artifacts.
+			if n.Outputs != nil {
+				for _, output := range n.Outputs.Parameters {
+					if !strings.HasPrefix(output.ValueFrom.Path, "/output/ml_metadata/") {
+						continue
+					}
+
+					artifacts, err := parseTFXMetadata(*output.Value)
+					if err != nil {
+						return util.NewInternalServerError(err, "metadata parsing failure")
+					}
+
+					if err := s.storeArtifacts(artifacts); err != nil {
+						return util.NewInternalServerError(err, "artifact storing failure")
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *Store) storeArtifacts(artifacts artifactStructs) error {
+	for _, a := range artifacts {
+		id, err := s.mlmdStore.PutArtifactType(
+			a.ArtifactType, &mlmetadata.PutTypeOptions{AllFieldsMustMatch: true})
+		if err != nil {
+			return util.NewInternalServerError(err, "failed to register artifact type")
```

Comment:
It's an internal error. It can't be an invalid input error, as that's for user input right?

Reply:
An internal error is when something is broken in either the API server or a main piece of infrastructure. 

So: 
- Internal error would be when a connection to the DB cannot be made. 
- If the data returned by the container is wrong, there is nothing wrong with the pipeline system itself, so it shouldn't be an internal error. (and clients shouldn't retry). 

In this case, it seems that we need to distinguish between the two cases, correct?

Reply:
Sure. What's the right error here? I'll change it to whatever you suggest is the right error type.

Reply:
I think we need two cases: 
- if the metadata content is the problem => InvalidInputError
- Otherwise (connection error, etc.) => InternalServerError.

Reply:
Done